### PR TITLE
Add a new separator for speech enhancement/separation tasks

### PR DESCRIPTION
**Pull request by LiChenda**
This PR is an implementation for [Skipping Memory (SkiM)](https://arxiv.org/abs/2201.10800) net. 
It is a low-computational-cost model for long sequence modeling.
In the [original paper](https://arxiv.org/abs/2201.10800), it is used for low-latency real-time Continuous Speech Separation (CSS).
Currently, the CSS task is not ready in the main ESPNet branch. So this PR mainly Implements the SkiM model. 
For testing the model itself, I also add two related configuration files (causal and non-causal) in the WSJ0-2mix recipe. (By using a small encoder kernel size, the utterance-level input also generates a long feature sequence.)

### CODE REVIEW

File: espnet2/enh/layers/tcn.py [hunk 1]
```diff
@@ -253,20 +258,37 @@ def forward(self, y):
         Returns:
             cLN_y: [M, N, K]
         """
+        dim = 3
+        if y.dim() == 4:
```
Comment (sw005320) [inline espnet2/enh/layers/tcn.py:262]: Can you add a test for this case?

**Review by sw005320 — COMMENTED** [review]
How about adding a result to wsj0_2mix/enh1/README.md with pre-trained models?
WSJ0_2mix is not a suitable example, but it is still useful.

**Comment by sw005320** [comment]
LGTM for me.
@Emrys365, do you have some comments?
I think this PR is already in a good shape.

**Comment by LiChenda** [comment]
> How about adding a result to wsj0_2mix/enh1/README.md with pre-trained models? WSJ0_2mix is not a suitable example, but it is still useful.

Yes, I have some pre-trained models on WSJ0_2mix. However, I rewrote the code for this open-source PR to avoid some IP issues, thus the previously trained model can't be directly loaded due to some renaming. Modifying the parameter keys in the previous checkpoint file might be too hacking. So, I am training a new model with this codebase. The progress looks good. I'll upload a pre-trained model when the training is done.

File: espnet2/enh/layers/skim.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+                "LSTM",
+                input_size=self.input_size,
+                hidden_size=self.hidden_size,
+                dropout=dropout,
+                bidirectional=bidirectional,
+            )
+            self.h_norm = chose_norm(
+                norm_type=norm_type, channel_size=self.input_size, shape="BTD"
+            )
+        if mem_type in ["hc", "c"]:
```
[the remark this answers is not in the document]
Comment (LiChenda) [reply espnet2/enh/layers/skim.py:63]: Yes, this _if_ is for the "hidden and cell" (hc) and "cell" (c) state.

File: espnet2/enh/layers/skim.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+        return output
+
+    def _padfeature(self, input):
+        B, T, D = input.shape
+        rest = self.segment_size - T % self.segment_size
+
+        if rest > 0:
+            pad = torch.zeros(B, rest, D, device=input.device)
+            input = torch.cat([input, pad], dim=1)
```
Comment (Emrys365) [inline espnet2/enh/layers/skim.py:298]: You can simplify the code with:
```python
            input = torch.nn.functional.pad(input, (0, 0, 0, rest))
```

**Review by Emrys365 — APPROVED** [review]
LGTM. I just added some minor comments.

File: espnet2/enh/layers/tcn.py (hunk 1 above)
Comment (LiChenda) [reply espnet2/enh/layers/tcn.py]: This if-branch is never used in the codebase, so I removed this case.

File: espnet2/enh/separator/skim_separator.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            others predicted data, e.g. masks: OrderedDict[
+                'mask_spk1': torch.Tensor(Batch, Frames, Freq),
+                'mask_spk2': torch.Tensor(Batch, Frames, Freq),
+                ...
+                'mask_spkn': torch.Tensor(Batch, Frames, Freq),
+            ]
+        """
+
+        # if complex spectrum,
+        if isinstance(input, ComplexTensor):
```
[the remark this answers is not in the document]
Comment (LiChenda) [reply espnet2/enh/separator/skim_separator.py]: Done.

File: espnet2/enh/layers/tcn.py [hunk 5]
```diff
@@ -281,9 +303,15 @@ def forward(self, y):
         Returns:
             gLN_y: [M, N, K]
         """
+        if self.shape == "BTD":
+            y = y.transpose(1, 2).contiguous()
+
         mean = y.mean(dim=1, keepdim=True).mean(dim=2, keepdim=True)  # [M, 1, 1]
```
[the remark this answers is not in the document]
Comment (LiChenda) [reply espnet2/enh/layers/tcn.py]: Done.

File: espnet2/enh/layers/tcn.py [hunk 6]
```diff
@@ -281,9 +303,15 @@ def forward(self, y):
...
         var = (
             (torch.pow(y - mean, 2)).mean(dim=1, keepdim=True).mean(dim=2, keepdim=True)
```
[the remark this answers is not in the document]
Comment (LiChenda) [reply espnet2/enh/layers/tcn.py]: Done.

File: espnet2/enh/layers/skim.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+# (https://arxiv.org/abs/2201.10800)
+#
+
+import torch
+import torch.nn as nn
+
+from espnet2.enh.layers.dprnn import merge_feature
+from espnet2.enh.layers.dprnn import SingleRNN
+from espnet2.enh.layers.dprnn import split_feature
+from espnet2.enh.layers.tcn import chose_norm
```
[the remark this answers is not in the document]
Comment (LiChenda) [reply espnet2/enh/layers/skim.py]: Done.

File: espnet2/enh/layers/skim.py (hunk 3 above)
Comment (LiChenda) [reply espnet2/enh/layers/skim.py]: Done.

**Comment by LiChenda** [comment]
The CI failure seems to be caused by a [pytorch_complex-related issue](https://github.com/kamo-naoyuki/pytorch_complex/pull/10). Maybe we can retry the CI test when the issue is solved.

**Pull request merged**